### PR TITLE
Dragonflight compatibility

### PR DESCRIPTION
**Pull request by padic-dev**
Alex,

  Not sure if you are still actively working to repair this add-on for DragonFlight. I have tried a few times to make this work for my needs and have been updating the issue with my changes. I was able to make a more viable version today and decided to make a pull request for you to review. 

  No promises it is perfectly working for 10.0.2 but basic gear equipping seems to be working for my needs.  I did add some comments where I made changes to help relay why I made the changes I did. Hopefully this helps you and the community who use your add-on. 

  I did not update the TOC file since this is not a complete fix and it is your add-on to make the decisions on. Also I did not test the changes for Classic or Wrath but I did try to use cross platform logic in most places as I understand it.

-Padic

### CODE REVIEW

**Comment by shot1330** [comment]
yeah Check this out Alex

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:2465]: We do not want this behavior.  It would prevent removing an item manually.  AutoGear would fight against the player immediately instead of allowing the player to move their items manually as they would wish.  AutoGear should only automatically equip an item under these circumstances:

- the player's character receives a gear item for the first time from outside their inventory, like from a loot roll, vendor, or a quest reward
- the player's character's gear evaluation situation changes, like when changing specialization
- the player has manually initiated a scan

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:2227]: We do not want the behavior to which the registration of this event leads.  It would prevent removing an item manually.  AutoGear would fight against the player immediately instead of allowing the player to move their items manually as they would wish.  AutoGear should only automatically equip an item under these circumstances:

- the player's character receives a gear item for the first time from outside their inventory, like from a loot roll, vendor, or a quest reward
- the player's character's gear evaluation situation changes, like when changing specialization
- the player has manually initiated a scan

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:3943]: We previously hooked 4 tooltips, so this should either check for all the previous tooltip types or check the new Enum for the tooltip data type `Item`.  There may be a way to check that.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:2425]: You could write it this way, which might actually be my preference for this kind of thing:

```Lua
if C_Container and C_Container.PickupContainerItem then
	C_Container.PickupContainerItem(i, j)
elseif PickupContainerItem then
	PickupContainerItem(i, j)
else
	AutoGearPrint("AutoGear: Can't pick up an item because function PickupContainerItem does not exist in this version of the WoW addon API ("..tostring(TOC_VERSION_CURRENT)..").  Returning.",0)
	return
end
```

This way, you're not relying on the return value of the ternary, and simply relying on the true or nil of the if conditional before executing the exact function you intend.  I'm not sure if there's a real difference, but it seems clearer to me and allows for additional logic like an error message and early return.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:4191]: I'm curious how this worked while the comment in a change above for this exact same function claims to not have worked.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
This is very helpful.  Thank you!  I haven't had sufficient interest for my limited free time for installing and configuring Dragonflight just for the purpose of fixing AutoGear, so I appreciate your effort tremendously.

I've reviewed your code changes and I'm requesting some changes.  I'd like to avoid changing the behavior (aside from not working, of course) and stick to compatibility fixes only.

Also, I'm requesting removal of the extra comments, as they have finished serving their purpose.  Thank you for leaving them initially, as they helped me understand the rationale for your changes.  However, we will not need them inside the Lua code from now on, and we can interact in github review comments instead.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:3943]: As per the API changes, it seems the previous 4 hooks are handled by the new single line. The first parameter of the new handler determines what type of tooltips to hook to and works globally. It was my understanding the one line now does the same as the old 4 hooks. I could be mistaken and I am also guessing here as to the intention of the addon as a whole.

Tooltip Script Handlers

The following GameTooltip script handlers have been removed. This change will not be reflected in the UI XML schema definition until a future patch.

[OnTooltipAddMoney](https://wowpedia.fandom.com/wiki/UIHANDLER_OnTooltipAddMoney)
OnTooltipSetAchievement
OnTooltipSetEquipmentSet
[OnTooltipSetItem](https://wowpedia.fandom.com/wiki/UIHANDLER_OnTooltipSetItem)
OnTooltipSetQuest
OnTooltipSetSpell
OnTooltipSetUnit
Usages of these script handlers should be replaced by registering callbacks with the new [TooltipDataProcessor.AddTooltipPostCall](https://github.com/Ketho/wow-ui-source-df/search?q=AddTooltipPostCall) function. Note that callbacks registered with this mechanism are global and will be triggered for all tooltips which inherit from GameTooltipTemplate.

local function OnTooltipSetItem(tooltip, data)
    if tooltip == GameTooltip then
        print("OnTooltipSetItem", tooltip, data)
    end
end

-- Replace 'Enum.TooltipDataType.Item' with an appropriate type for the tooltip
-- data you are wanting to process; eg. use 'Enum.TooltipDataType.Spell' for
-- replacing usage of OnTooltipSetSpell.
--
-- If you wish to respond to all tooltip data updates, you can instead replace
-- the enum with 'TooltipDataProcessor.AllTypes' (or the string "ALL").

TooltipDataProcessor.AddTooltipPostCall(Enum.TooltipDataType.Item, OnTooltipSetItem)

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:3948]: This can be changed back, it was a proactive guess since I was having issues with the Tooltip:GetItem() call as well, I made change thinking it would help with the issues.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:2227]: See previous comment, I changed this locally to listen to BAG_NEW_ITEMS_UPDATED

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:3943]: What I mean is that we want AutoGear to operate on the ShoppingTooltip1, ShoppingTooltip2, and ItemRefTooltip, if those still exist at all.  The logic in the line upon which this review comment thread is based excludes those tooltips specifically by checking for `GameTooltip` only, instead of checking `Enum.TooltipDataType.Item`, if that is possible to check.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:4191]: You and me both, it was somewhat frustrating that the above one that didn't work was the only one. It kept throwing an error

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:2425]: I can make that change

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:2465]: There is a useful programming tool built into the WoW interface which can listen for events, accessible with the chat command `/eventtrace`.  If you try using that, then getting a new item, you may see the necessary event.

Currently, before any of these changes, AutoGear reads chat messages and uses a specific filter to find the item link in the chat message.  It seems likely that if Blizzard changed those chat messages, the filter may no longer match, so my advice is to start there, after confirming that the chat message event is at least firing.

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:4191]: BugGrabber often groups identical errors together, even if they occur in different parts of the code.  It's possible both parts were throwing the error, but the stack trace only showed the first one, if you were looking at BugGrabber.  Alternatively, it's also possible that your testing only hit the other code block, so this part was not executed.

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:3948]: Sounds good.

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:2227]: OK.  My other comment touches on the most likely reason the auto-equipping "stopped working": the chat filter no longer matching.  You may want to start there, and/or make use of `/eventtrace` to find exactly what event is thrown with which arguments when receiving new gear from outside the player's inventory.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Thanks for your commentary.  I've added some replies.  Please feel free to commit to your branch with your changes when they're made.  I'm just continuing this review in order to add my comment replies.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:2465]: Yeah, I stumbled upon /etrace which is how I found the bag update events. I did also notice the chat parsing logic today when looking over things again. I plan to look at that further and see if I can figure out what broke there to minimize the overall changes to AutoGear.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:3943]: I did some digging and found lua and xml code from the 10.0.2 beta which shows GameTooltip and ItemRefTooltip inherits from GameTooltipTemplate so the new hook should cover that. ShoppingTooltip* inherit from ShoppingTooltipTemplate which I can't find anything on so I plan to try to return those previous hooks to see what that does.

https://github.com/tomrus88/BlizzardInterfaceCode/blob/168d22d05fdaf7d3ca2e1af0d02a5cd9707b15b5/Interface/FrameXML/GameTooltip.xml

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:4191]: I am using BugGrabber and I was thinking the same, possibly I only hit the one code block that was erroring. I am planning to dig into that a little more.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:2465]: In case this helps you see something I am missing with less time involved than installing the game. If not no biggie, I tend to be verbose and over share in case it might be helpful.

Screenshot 2022-12-12 at 8 32 10 PM

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:3943]: I think I finally understand what you were getting at. I tried to add the ShoppingTooltip* hooks and that threw an error, that was when I realized the issue wasn't the new hook but my if check. I updated that for my code and will have that with my next checkin.

if tooltip ~= GameTooltip or tooltip ~= ShoppingTooltip1 or tooltip ~= ShoppingTooltip2 or tooltip ~= ItemRefTooltip or tooltip == nil then return end

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:2465]: The issue seems to be the event passes the unit name with -Server appended, UnitName("Player") does not causing the listener for chat messages to break out early. Looking for the best fix for that now.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:3948]: This logic does not make sense.  If the tooltip is `ShoppingTooltip1`, the first condition, `tooltip ~= GameTooltip` will immediately evaluate as true, resulting in executing `return`.  You need `and` there.

This makes me concerned that this change wasn't tested.  The shopping tooltips are the comparison tooltips when hovering gear that is not equipped.

By the way, are you sure we can't instead simply check for whether the tooltip data type is `Enum.TooltipDataType.Item`?  I imagine that if they added that enum, they would have also added a way to check it.

Come to think of it, the hook itself  (this line: `TooltipDataProcessor.AddTooltipPostCall(Enum.TooltipDataType.Item, AutoGearTooltipHook)`) seems to exclude tooltips which are not of tooltip data type `Enum.TooltipDataType.Item`, so what makes this additional check necessary at all?  It did not exist in the original code, since the original code only hooked the tooltips it needed to, and the new code seems to also do that.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:4122]: This needs to support Classic clients still.  It should first check for the existence of `TooltipDataProcessor` and `TooltipDataProcessor.AddTooltipPostCall`, and if those don't exist, use the previous hooks, like this:

```Lua
if TooltipDataProcessor and TooltipDataProcessor.AddTooltipPostCall then
	TooltipDataProcessor.AddTooltipPostCall(Enum.TooltipDataType.Item, AutoGearTooltipHook)
else
	GameTooltip:HookScript("OnTooltipSetItem", AutoGearTooltipHook)
	ShoppingTooltip1:HookScript("OnTooltipSetItem", AutoGearTooltipHook)
	ShoppingTooltip2:HookScript("OnTooltipSetItem", AutoGearTooltipHook)
	ItemRefTooltip:HookScript("OnTooltipSetItem", AutoGearTooltipHook)
end
```

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:4195]: I see you've now updated the other instance of this code, but not this instance.  If the other instance needed to be updated because it caused an error, it's very likely this one needs to be updated as well.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Thanks for these changes!  It looks like this is almost ready for merging, but I still have some concerns, the most major being that your changes would certainly break Wrath Classic and all shopping tooltips.  I've detailed these concerns in this review.

File: AutoGear.lua
Comment (padic-dev) [reply AutoGear.lua:4122]: I simply attempted to fix the code the best I could. Leaving the older HookScript calls threw errors in Dragonflight which is where I am focused on playing currently. Hopefully someone else can dig into the Wrath issues and tooltip issues here. It seems to be working for the most part for my needs in Dragonflight.  Unfortunately I am not really available to dig into these issues further. Hopefully this will help move the add-on further for the current API.

Comment (AlexFolland) [reply AutoGear.lua:4122]: Even with my example, it didn't work?  If that's the case, Dragonflight is seriously messed up.

Comment (padic-dev) [reply AutoGear.lua:4122]: I did not code your example. The changes I made are working for my needs. Not to sound like a jerk, I offered my changes as help to move this forward, not as your coder for your add-on which you are not actively coding for. I provided my changes to help not to be someone to do your code changes for you. If this is the change you are curious about feel free to install Dragonflight and Wrath classic and test it out. I am quite busy now with the holidays in a few days and don't have time between my full-time job, my family and the holidays to do this coding when you yourself admitted you don't have time. It's beginning to feel like you are expecting me to be your code monkey so you don't have to be bothered with maintaining your add-on. 

I was happy to help because I was able to get this add-on working for myself and no one else had made any progress for quite awhile after the pre-patch release. I don't have the time nor interest in being the guy to do the grunt work when it is not even my add-on and the owner doesn't have time to chip in either. Sorry if this sounds rude, I just wanted to contribute but each time I offer my changes, I seem to get back, well that isn't what I want, how about you do this instead.

Comment (AlexFolland) [reply AutoGear.lua:4122]: Yes, what you've said is considered rude in a pull request comment on github generally, but I understand that you aren't familiar with the process of pull requests and code reviews, so I don't have a problem with what you've said.  I'll help you understand.

This is how a pull request (or if on github, a merge request) works.  You submit changes you'd like merged, then if approved, they are merged.  If not approved, there is feedback explaining why.  We, as programmers, should not become upset by the process.  We are all volunteering our time, and we may have any number of other more prevalent interests for which to use our time.

In this particular case, your fork is extremely close to working well.  However, it's obvious to me that the tooltips would not even work in Dragonflight, let alone Wrath Classic, because of the additional check on which I commented, so I can't merge the changes as they are.  Since the addon still works in its supported environment, I will continue awaiting the 2 small fixes necessary before the merge.  I'd like to help you understand why those 2 small changes are necessary, as well, which is why I spent my time reviewing and commenting.  Just as you are more interested in Dragonflight, I am more interested in Wrath Classic, and I don't feel a particular need to merge unfinished and/or untested changes into a code base which is currently working.

Please don't imply that I am doing something wrong by following the usual github pull request process.  If you don't like upstream, you are free to continue to use your own fork of the project without resolving any issues found in your pull request's code review.  As an author of the original project, I allow this, and the project's license reflects that.

**Comment by AlexFolland** [comment]
I made some changes for compatibility with the current code base and to maintain Wrath Classic compatibility.  I tested in-game in Wrath Classic and it seemed to work, so I pushed and tagged the release.  Thanks for your effort with this.  Ultimately, it came down to 2 necessary changes, along with some changes that were contributed by a different contributor.  Would you mind testing this in Dragonflight?

**Comment by AlexFolland** [comment]
It seems to have exhibited errors immediately, unfortunately, so I've undone it for now until the tooltip issue can be resolved.  More info is in #40.

**Pull request merged**
